File: COMPASS/survival_analysis/prepare_metastatic_figure_labels.py
"""Small label bridge for 05_figures.Rmd; never reads longitudinal lab data.

Reuse the pipeline's medication-derived ADT intent labels, and compare them
with the metastatic-diagnosis LLM task and dated regex stages. Stage I–III is
local; stage IV is metastatic. Missing/invalid values remain unclassified.
"""
from __future__ import annotations

import argparse
from pathlib import Path
import sys

import polars as pl

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

ID = "DFCI_MRN"
LOCAL = "Local"
MET = "Metastatic"


def normalize_id(frame: pl.DataFrame) -> pl.DataFrame:
    return frame.with_columns(
        pl.col(ID).cast(pl.Float64, strict=False).cast(pl.Int64, strict=False)
    ).drop_nulls(ID)


def collapse_stage(column: str) -> pl.Expr:
    stage = pl.col(column).cast(pl.String).str.strip_chars().str.to_uppercase()
    return (
        pl.when(stage.is_in(["1", "2", "3", "1.0", "2.0", "3.0", "I", "II", "III"]))
        .then(pl.lit(LOCAL))
        .when(stage.is_in(["4", "4.0", "IV"]))
        .then(pl.lit(MET))
        .otherwise(None)
    )


def build_labels(intent: pl.DataFrame, notes: pl.DataFrame,
                 llm: pl.DataFrame) -> pl.DataFrame:
    intent = normalize_id(intent).with_columns(
        pl.col("ADT_FIRST_DATE").cast(pl.String).str.to_datetime(strict=False),
        pl.col("ADT_INTENT").replace_strict(
            {"LOCALIZED_ADJUVANT": LOCAL, "METASTATIC": MET}, default=None
        ).alias("ADT_LABEL"),
    )
    if intent[ID].n_unique() != intent.height:
        raise ValueError("ADT intent input must contain exactly one row per patient")

    # The input is the metastatic-diagnosis task, not the NEPC subtype task.
    verdict = pl.col("has_metastatic_disease").cast(pl.String).str.to_lowercase()
    llm = normalize_id(llm).with_columns(
        pl.when(verdict.is_in(["true", "1", "1.0"])).then(pl.lit(MET))
        .when(verdict.is_in(["false", "0", "0.0"])).then(pl.lit(LOCAL))
        .otherwise(None).alias("LLM_LABEL")
    ).group_by(ID).agg(
        # Match compass_pipeline: any positive wins, missing stays missing.
        pl.when((pl.col("LLM_LABEL") == MET).any()).then(pl.lit(MET))
        .when((pl.col("LLM_LABEL") == LOCAL).any()).then(pl.lit(LOCAL))
        .otherwise(None).alias("LLM_LABEL")
    )
    stage_text = pl.col("DERIVED_STAGE_MERGED").cast(pl.String).str.strip_chars().str.to_uppercase()
    notes = normalize_id(notes).select(
        ID,
        pl.col("EVENT_DATE").cast(pl.String).str.to_datetime(strict=False).alias("stage_date"),
        stage_text.replace_strict(
            {"1": 1, "2": 2, "3": 3, "4": 4,
             "I": 1, "II": 2, "III": 3, "IV": 4,
             "1.0": 1, "2.0": 2, "3.0": 3, "4.0": 4}, default=None,
        ).alias("stage"),
    ).drop_nulls(["stage_date", "stage"])
    joined = notes.join(intent.select(ID, "ADT_FIRST_DATE"), on=ID).with_columns(
        (pl.col("stage_date") - pl.col("ADT_FIRST_DATE")).dt.total_days().alias("days")
    )
    before = joined.filter(pl.col("days") <= 0)
    nearest = before.filter(pl.col("days") >= -365).sort(
        ["days", "stage"], descending=[True, True]
    ).group_by(ID).agg(pl.col("stage").first().alias("REGEX_STAGE"))
    result = intent.join(llm, on=ID, how="left").join(nearest, on=ID, how="left")
    for suffix, frame in [("BEFORE", before), ("AFTER", joined.filter(pl.col("days") > 0))]:
        result = result.join(frame.group_by(ID).agg(
            pl.col("stage").max().alias(f"REGEX_MAX_{suffix}_STAGE")
        ), on=ID, how="left")
    return result.with_columns(
        collapse_stage("REGEX_STAGE").alias("REGEX_LABEL"),
        collapse_stage("REGEX_MAX_BEFORE_STAGE").alias("REGEX_MAX_BEFORE"),
        collapse_stage("REGEX_MAX_AFTER_STAGE").alias("REGEX_MAX_AFTER"),
    )


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--intent", required=True)
    parser.add_argument("--stage", required=True)
    parser.add_argument("--llm", required=True)
    parser.add_argument("--icd")
    parser.add_argument("--output", required=True)
    args = parser.parse_args()
    intent = pl.read_csv(args.intent, infer_schema_length=0)
    notes = pl.read_parquet(args.stage, columns=[ID, "EVENT_DATE", "DERIVED_STAGE_MERGED"])
    llm = pl.read_parquet(args.llm, columns=[ID, "has_metastatic_disease"])
    labels = build_labels(intent, notes, llm)
    if args.icd and Path(args.icd).exists():
        from COMPASS.data_preprocessing.validate_adt_intent import compute_met_burden_at_adt
        icds = pl.read_csv(args.icd, infer_schema_length=0,
                           columns=[ID, "DIAGNOSIS_ICD10_CD", "START_DT"])
        burden = compute_met_burden_at_adt(icds, labels)
        labels = labels.join(burden, on=ID, how="left")
    # Only the fields used by R cross the bridge. R restricts to the canonical
    # time-zero analysis cohort before computing any summaries.
    columns = [ID, "ADT_FIRST_DATE", "ADT_LABEL", "LLM_LABEL", "REGEX_LABEL",
               "REGEX_MAX_BEFORE", "REGEX_MAX_AFTER"]
    columns += [c for c in labels.columns if c == "N_MET_SITES" or c.startswith("MET_SITE_")]
    labels.select(columns).write_csv(args.output)


if __name__ == "__main__":
    main()

File: COMPASS/survival_analysis/prepare_metastatic_figure_labels.R
# R-only preparation for 05_figures.Rmd. Uses the existing intent labels,
# metastatic-diagnosis LLM task and dated regex stages; no model refits.
metastatic_parquet_backend <- function() {
  if (requireNamespace("arrow", quietly = TRUE)) return("arrow")
  if (requireNamespace("nanoparquet", quietly = TRUE)) return("nanoparquet")
  stop('Metastatic figures need an R Parquet reader. Run install.packages("nanoparquet") ',
       'in this R environment, then re-knit. The R package arrow is also supported.', call. = FALSE)
}

read_metastatic_parquet <- function(path, columns) {
  if (metastatic_parquet_backend() == "arrow")
    return(as.data.frame(arrow::read_parquet(path, col_select = columns)))
  nanoparquet::read_parquet(path, col_select = columns)
}

metastatic_normalize_id <- function(frame) {
  id <- suppressWarnings(as.numeric(as.character(frame$DFCI_MRN)))
  keep <- is.finite(id)
  frame <- frame[keep, , drop = FALSE]
  frame$DFCI_MRN <- format(trunc(id[keep]), scientific = FALSE, trim = TRUE)
  frame
}

metastatic_datetime <- function(values) {
  if (inherits(values, "POSIXt")) return(as.POSIXct(values, tz = "UTC"))
  if (inherits(values, "Date")) return(as.POSIXct(values, tz = "UTC"))
  values <- trimws(as.character(values))
  parsed <- suppressWarnings(readr::parse_datetime(values))
  for (format in c("%Y-%m-%d %H:%M:%S", "%Y-%m-%d", "%m/%d/%Y %H:%M:%S", "%m/%d/%Y")) {
    missing <- is.na(parsed) & !is.na(values) & nzchar(values)
    if (!any(missing)) break
    parsed[missing] <- suppressWarnings(readr::parse_datetime(values[missing], format = format))
  }
  parsed
}

metastatic_stage_number <- function(values) {
  key <- toupper(trimws(as.character(values)))
  lookup <- c(`1`=1L, `2`=2L, `3`=3L, `4`=4L, I=1L, II=2L, III=3L, IV=4L,
              `1.0`=1L, `2.0`=2L, `3.0`=3L, `4.0`=4L)
  unname(lookup[key])
}

metastatic_collapse_stage <- function(values) {
  stage <- metastatic_stage_number(values)
  ifelse(is.na(stage), NA_character_, ifelse(stage == 4L, "Metastatic", "Local"))
}

build_metastatic_labels <- function(intent, notes, llm, analysis_anchors = NULL) {
  intent <- metastatic_normalize_id(intent) %>%
    transmute(DFCI_MRN, ADT_FIRST_DATE = metastatic_datetime(ADT_FIRST_DATE),
      ADT_LABEL = unname(c(LOCALIZED_ADJUVANT = "Local", METASTATIC = "Metastatic")[as.character(ADT_INTENT)]))
  if (anyDuplicated(intent$DFCI_MRN)) stop("ADT intent input must contain exactly one row per patient")
  if (is.null(analysis_anchors)) {
    # Standalone use without an analysis cohort retains the intent-file clock.
    intent <- intent %>% mutate(ANALYSIS_ANCHOR_DATE = ADT_FIRST_DATE)
  } else {
    if (!all(c("DFCI_MRN", "TREATMENT_ANCHOR_DATE") %in% names(analysis_anchors)))
      stop("Metastatic supplement requires patient DFCI_MRN and TREATMENT_ANCHOR_DATE")
    anchors <- metastatic_normalize_id(analysis_anchors) %>%
      transmute(DFCI_MRN, ANALYSIS_ANCHOR_DATE = metastatic_datetime(TREATMENT_ANCHOR_DATE))
    if (anyDuplicated(anchors$DFCI_MRN)) stop("Analysis anchors must contain exactly one row per patient")
    # Match by patient ID, never row order. Preserve patients without an intent
    # label so their independent LLM and regex evidence is still represented.
    intent <- anchors %>% left_join(intent, by = "DFCI_MRN")
  }
  intent <- intent %>% mutate(ANCHOR_DELTA_DAYS = as.numeric(
    as.Date(ANALYSIS_ANCHOR_DATE, tz = "UTC") - as.Date(ADT_FIRST_DATE, tz = "UTC")))
  llm <- metastatic_normalize_id(llm) %>%
    mutate(.verdict = tolower(as.character(has_metastatic_disease))) %>%
    group_by(DFCI_MRN) %>% summarise(LLM_LABEL = case_when(
      any(.verdict %in% c("true", "1", "1.0")) ~ "Metastatic",
      any(.verdict %in% c("false", "0", "0.0")) ~ "Local",
      TRUE ~ NA_character_), .groups = "drop")
  notes <- metastatic_normalize_id(notes) %>%
    transmute(DFCI_MRN, .date = metastatic_datetime(EVENT_DATE),
              .stage = metastatic_stage_number(DERIVED_STAGE_MERGED)) %>%
    filter(!is.na(.date), !is.na(.stage)) %>%
    inner_join(intent %>% select(DFCI_MRN, ANALYSIS_ANCHOR_DATE), by = "DFCI_MRN") %>%
    mutate(.days = trunc(as.numeric(difftime(.date, ANALYSIS_ANCHOR_DATE, units = "days"))))
  before <- notes %>% filter(.days <= 0)
  nearest <- before %>% filter(.days >= -365) %>% arrange(desc(.days), desc(.stage)) %>%
    distinct(DFCI_MRN, .keep_all = TRUE) %>% transmute(DFCI_MRN, REGEX_STAGE = .stage)
  max_before <- before %>% group_by(DFCI_MRN) %>%
    summarise(REGEX_MAX_BEFORE_STAGE = if (length(.stage)) max(.stage) else NA_integer_, .groups = "drop")
  max_after <- notes %>% filter(.days > 0) %>% group_by(DFCI_MRN) %>%
    summarise(REGEX_MAX_AFTER_STAGE = if (length(.stage)) max(.stage) else NA_integer_, .groups = "drop")
  # Maximum across the entire record: pre- and post-ADT staging pooled, with no
  # anchor window. Post-ADT progression therefore counts toward the label, so
  # this is not a point-in-time status at ADT the way REGEX_LABEL is.
  max_any <- notes %>% group_by(DFCI_MRN) %>%
    summarise(REGEX_MAX_ANY_STAGE = if (length(.stage)) max(.stage) else NA_integer_, .groups = "drop")
  intent %>% left_join(llm, by = "DFCI_MRN") %>% left_join(nearest, by = "DFCI_MRN") %>%
    left_join(max_before, by = "DFCI_MRN") %>% left_join(max_after, by = "DFCI_MRN") %>%
    left_join(max_any, by = "DFCI_MRN") %>%
    mutate(REGEX_LABEL = metastatic_collapse_stage(REGEX_STAGE),
           REGEX_MAX_BEFORE = metastatic_collapse_stage(REGEX_MAX_BEFORE_STAGE),
           REGEX_MAX_AFTER = metastatic_collapse_stage(REGEX_MAX_AFTER_STAGE),
           REGEX_MAX_ANY = metastatic_collapse_stage(REGEX_MAX_ANY_STAGE)) %>%
    select(DFCI_MRN, ADT_FIRST_DATE, ANALYSIS_ANCHOR_DATE, ANCHOR_DELTA_DAYS,
           ADT_LABEL, LLM_LABEL, REGEX_LABEL, REGEX_MAX_BEFORE, REGEX_MAX_AFTER,
           REGEX_MAX_ANY)
}

metastatic_icd_site <- function(values) {
  code <- gsub("[^A-Z0-9]", "", toupper(trimws(as.character(values))))
  valid <- !is.na(code) & nchar(code) >= 4L & grepl("^C7[789]", code) & code != "C799"
  site <- rep(NA_character_, length(code))
  site[valid] <- "other"
  # Mirrors validate_adt_intent.py: spinal/cranial-nerve codes remain "other";
  # unspecified C79.9 and C7B neuroendocrine codes do not count as organ sites.
  prefixes <- c(C795="bone", C7931="brain", C7932="brain", C797="adrenal",
                C780="lung", C782="lung", C787="liver", C786="peritoneal", C77="node")
  for (prefix in names(prefixes)) site[valid & startsWith(code, prefix)] <- prefixes[[prefix]]
  site
}

metastatic_burden_at_adt <- function(icds, labels) {
  groups <- c("brain", "bone", "liver", "lung", "node", "adrenal", "peritoneal", "other")
  anchors <- labels %>% select(DFCI_MRN, ANALYSIS_ANCHOR_DATE) %>% filter(!is.na(ANALYSIS_ANCHOR_DATE))
  coded <- metastatic_normalize_id(icds) %>% inner_join(anchors, by = "DFCI_MRN") %>%
    mutate(.date = metastatic_datetime(START_DT), .site = metastatic_icd_site(DIAGNOSIS_ICD10_CD)) %>%
    filter(!is.na(.date), !is.na(.site), .date <= ANALYSIS_ANCHOR_DATE) %>% distinct(DFCI_MRN, .site)
  base <- labels %>% distinct(DFCI_MRN)
  for (group in groups)
    base[[paste0("MET_SITE_", group)]] <- as.integer(base$DFCI_MRN %in% coded$DFCI_MRN[coded$.site == group])
  # With no analysis date, "before ADT" is unknown rather than zero burden.
  unavailable <- !base$DFCI_MRN %in% anchors$DFCI_MRN
  for (column in paste0("MET_SITE_", groups)) base[[column]][unavailable] <- NA_integer_
  base$N_MET_SITES <- rowSums(base[paste0("MET_SITE_", groups)])
  base
}

prepare_metastatic_figure_labels <- function(config, analysis_anchors = NULL) {
  intent <- readr::read_csv(config$intent, show_col_types = FALSE,
    col_select = all_of(c("DFCI_MRN", "ADT_FIRST_DATE", "ADT_INTENT")),
    col_types = readr::cols(.default = readr::col_character()))
  notes <- read_metastatic_parquet(config$stage, c("DFCI_MRN", "EVENT_DATE", "DERIVED_STAGE_MERGED"))
  llm <- read_metastatic_parquet(config$llm, c("DFCI_MRN", "has_metastatic_disease"))
  labels <- build_metastatic_labels(intent, notes, llm, analysis_anchors)
  if (!is.null(config$icd) && file.exists(config$icd)) {
    icds <- readr::read_csv(config$icd, show_col_types = FALSE,
      col_select = all_of(c("DFCI_MRN", "DIAGNOSIS_ICD10_CD", "START_DT")),
      col_types = readr::cols(.default = readr::col_character()))
    labels <- labels %>% left_join(metastatic_burden_at_adt(icds, labels), by = "DFCI_MRN")
  }
  labels
}
